documentation/wiki/decisions/adr/2024-10-18-document-storage.md

```diff
@@ -0,0 +1,93 @@
+# Simpler needs to store and lifecycle documents associated with Opportunities, which we will do via AWS S3 buckets
+
+- **Status:** Active<!-- REQUIRED -->
+- **Last Modified:** 2024-10-18 <!-- REQUIRED -->
+- **Related Issue:** [#2277](https://github.com/HHS/simpler-grants-gov/issues/2277) <!-- RECOMMENDED -->
+- **Deciders:** Matt Dragon, Aaron Couch, Kai Siren, Michael Chouinard, Lucas Brown<!-- REQUIRED -->
+- **Tags:** nofo, document, attachment, s3, storage <!-- OPTIONAL -->
+
+## Context and Problem Statement
+
+Opportunities include supporting documents that help define the opportunity, provide more instructions about applying, or otherwise supplement the Opportunity Listing. These documents represent individual files, sometimes within a folder/directory hierarchy that are provided to the Grant Seekers as a single Zip download currently. Among these files is one very special file, the Notice of Funding Opportunity (NOFO) that every Opportunity must publish.
+
+## Decision Drivers <!-- RECOMMENDED -->
+
+- Use the platform
+- Minimize cost per file (there will be a large number of files, but most will be rarely, if ever, accessed once the Opportunity closes)
+- Ease of processing - Use the best tools, with already supported libraries as they're intended
+- Follow best practices
+
+## Options Considered
+
+- [AWS Simple Storage Service (S3)](#aws-simple-storage-service-s3-buckets) bucket(s)
+- [Store files in PostgreSQL](#store-files-in-postgresql)
+- [Other off-the-shelf or homegrown storage solution](#other-off-the-shelf-or-homegrown-storage-solution)
+
+## Decision Outcome <!-- REQUIRED -->
+
+Chosen option: "AWS S3", because it represents the lowest Total Cost of Ownership (TCO) and industry best practices, including baked in support for access control for files, backups, etc. We will use 2 buckets, one for Published Opportunities, and one for DRAFT Opportunities. When an Opportunity is Published we'll ensure that the associated documents are copied to the Published bucket, making them accessible to the general public. Prior to publishing the documents will be accounted for in S3, so that the file storage is consistent throughout the lifecycle, but only the Publishing Service will have access to the files, ensuring they are not released to the public before the Opportunity and they can be revoked from public view if the Opportunity is accidentally Published.
```

Comment:
> We will use 2 buckets, one for Published Opportunities, and one for DRAFT Opportunities.

Are we planning on having the Published Opportunities bucket be publically accessible? Like truly 100% public?

Reply:
They'll be behind a CDN, so not "directly" publicly accessible, but close enough.

In the short-term before we have a CDN, we'll likely just use presigned s3 URLs which is a bit more direct, but we also dictate what is viewable in the API.

Reply:
To the best of my knowledge we don't need two buckets if the distribution method is cloudfront. cloudfront can [restrict distribution to a path](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/distribution-web-values-specify.html#DownloadDistValuesOriginPath) (say `published/`) and (this part I'm fuzzy on) [sign the requests to the bucket](https://aws.amazon.com/blogs/networking-and-content-delivery/amazon-cloudfront-introduces-origin-access-control-oac/).

I'm trying to do my diligence here since S3 bucket permissions are a massive thing for security people

Reply:
[I asked about this 🔒](https://nava.slack.com/archives/C03G1SWD9H7/p1729614170609279)

Reply:
Folks who hang out in platform are saying separate buckets so far! So I rescind mine opinion about using one bucket